### PR TITLE
Acdc seg

### PR DESCRIPTION
**Pull request by peter850706**

### CODE REVIEW

File: configs/train/acdc_adapt_from_scratch_config.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+logger:
+    name: 'AcdcAdaptLogger'
+
+monitor:
+    name: 'Monitor'
+    kwargs:
+        mode: 'min'
+        target: 'loss'
+        saved_freq: 1000
+        valid_freq: 10001
```
Comment (vic85821) [inline configs/train/acdc_adapt_from_scratch_config.yaml]: Set `valid_freq1` so as to avoid validating during the training process?

File: src/runner/trainers/acdc_seg_trainer.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            'loss': loss,
+            'losses': {
+                'CrossEntropyLoss': cross_entropy_loss,
+                'DiceLoss': dice_loss
+            }
+        }
+
+    def _valid_step(self, batch):
+        input, target = batch['input'].to(self.device), batch['target'].to(self.device)
+        output = F.interpolate(self.net(input),
```
Comment (vic85821) [inline src/runner/trainers/acdc_seg_trainer.py:30]: Better than dynamic upsampling in the network?

File: configs/train/acdc_adapt_from_scratch_config.yaml (hunk 1 above)
Comment (peter850706) [reply configs/train/acdc_adapt_from_scratch_config.yaml]: Yes.

File: src/runner/trainers/acdc_seg_trainer.py (hunk 2 above)
Comment (peter850706) [reply src/runner/trainers/acdc_seg_trainer.py:30]: I'm not sure what is "dynamic upsampling" you mentioned. Besides, I haven't tried other solution because the results are reasonable.

Comment (vic85821) [reply src/runner/trainers/acdc_seg_trainer.py:30]: `dynamic upsampling` means we dynamically adjust the shape of feature maps before they are concatenated by the skip connection.

Comment (peter850706) [reply src/runner/trainers/acdc_seg_trainer.py:30]: Sorry I'm quite confused. Any reference for this?

File: src/preprocess/acdc_data_split.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        with open(data_path, 'r') as f:
+            for line in f:
+                key, value = line.rstrip('\n').split(': ')
+                if key == 'Group':
+                    group_dict[value].append(data_path.parent.resolve())
+                    break
+    _, groups = zip(*sorted(group_dict.items()))
+    random.seed(0)
+    groups = tuple(random.sample(group, k=len(group)) for group in groups)
+    folds = tuple(zip(*groups))  # twenty folds where a fold contains five types of patients.
```
Comment (vic85821) [inline src/preprocess/acdc_data_split.py]: The comment may be a little ambiguous.

There are twenty folds because there are twenty cases in each group. Hence, we can state clearly for those readers who are unfamiliar with the ACDC dataset details.

File: src/preprocess/acdc_data_split.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    for data_path in data_paths:
+        with open(data_path, 'r') as f:
+            for line in f:
+                key, value = line.rstrip('\n').split(': ')
+                if key == 'Group':
+                    group_dict[value].append(data_path.parent.resolve())
+                    break
+    _, groups = zip(*sorted(group_dict.items()))
+    random.seed(0)
+    groups = tuple(random.sample(group, k=len(group)) for group in groups)
```
Comment (vic85821) [inline src/preprocess/acdc_data_split.py:28]: Does it aim to randomly shuffle the `data_paths` in each group?

File: src/preprocess/acdc_data_split.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    folds = tuple(zip(*groups))  # twenty folds where a fold contains five types of patients.
+
+    output_dir = args.output_dir
+    if not output_dir.is_dir():
+        output_dir.mkdir(parents=True)
+    ratio = len(folds) // (args.k + 1)
```
Comment (vic85821) [inline src/preprocess/acdc_data_split.py:37]: If `args.k` is 5, the dataset will be split into 6 folds (5 for training, 1 for testing)?

File: src/runner/trainers/acdc_seg_trainer.py (hunk 2 above)
Comment (vic85821) [reply src/runner/trainers/acdc_seg_trainer.py:30]: image

If using the `_UpBlock` in the network, why do we need to interpolate the output of the network into the shape of the target? The shapes of the output and the target should be the same, right?

Comment (peter850706) [reply src/runner/trainers/acdc_seg_trainer.py:30]: Actually it's not because the input data are isotropic (1mm x 1mm x 1mm) while the target is the original resolution. That's why we need to interpolate the network output.

File: src/preprocess/acdc_data_split.py (hunk 3 above)
Comment (peter850706) [reply src/preprocess/acdc_data_split.py]: Actually I just want to emphasize that a fold contains five types of patients (means we split evenly), not "there are twenty folds". I can add why there are twenty folds though.

File: src/preprocess/acdc_data_split.py (hunk 4 above)
Comment (peter850706) [reply src/preprocess/acdc_data_split.py:28]: Yes it's.

File: src/runner/trainers/acdc_seg_trainer.py (hunk 2 above)
Comment (vic85821) [reply src/runner/trainers/acdc_seg_trainer.py:30]: Got it. I misunderstand.

File: src/preprocess/acdc_data_split.py (hunk 5 above)
Comment (peter850706) [reply src/preprocess/acdc_data_split.py:37]: That's right.

File: src/preprocess/acdc_data_split.py (hunk 3 above)
Comment (vic85821) [reply src/preprocess/acdc_data_split.py]: OK!

File: src/preprocess/acdc_data_split.py (hunk 5 above)
Comment (vic85821) [reply src/preprocess/acdc_data_split.py:37]: In our experiment, we will use the `test_fold` instead of the cases in the testing directory for the verification of our algorithm, right?

File: src/preprocess/acdc_data_split.py (hunk 3 above)
Comment (peter850706) [reply src/preprocess/acdc_data_split.py]: Solved in 258d296.

File: src/preprocess/acdc_data_split.py (hunk 5 above)
Comment (peter850706) [reply src/preprocess/acdc_data_split.py:37]: Yes. But `testing.csv` can still be used if we can upload the results on their website.

File: src/data/datasets/acdc_adapt_dataset.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            patient_dirs = map(
+                Path,
+                data_split_file[
+                    (data_split_file.type == 'train') | (data_split_file.type == 'valid')
+                ].path
+            )
+            self.data_paths = tuple(
+                data_path
+                for patient_dir in patient_dirs
+                for data_path in sorted(patient_dir.glob('**/*frame??.nii.gz'))
```
Comment (vic85821) [inline src/data/datasets/acdc_adapt_dataset.py:37]: Can we use each frame in the 4D MRI data instead of only ED and ES two frames when we pre-train/adapt the network with ACDC dataset?

Comment (peter850706) [reply src/data/datasets/acdc_adapt_dataset.py:37]: I think it's a good point in practice; however, we want to claim/verify that it's our adapting strategy works, not because of the more data we use. So I think we should only use ED and ES frame.

Comment (vic85821) [reply src/data/datasets/acdc_adapt_dataset.py:37]: Got it.
